README.md
# fast-tldr

A fast client for tldr.

We strive to be the fastest tldr-client with no compromises regarding features.

This project is a fork of [tldr-hs](https://github.com/psibi/tldr-hs/).

## Features

- Cache pages
- Configurable platform and language
- 

## Installation

See Github releases: https://github.com/gutjuri/fast-tldr/releases

Or

1. [Install stack](https://docs.haskellstack.org/en/stable/README/#how-to-install)
2. `git clone https://github.com/gutjuri/fast-tldr.git && cd fast-tldr`
2. `stack install` or `stack install --local-bin-path .` 

## Usage

``` shellsession
$ tldr --help
Usage:
        tldr [options] COMMAND
        tldr [options]
Options:
-h --help                  Show help
-v,--version               Show version
-u --update                Update local cache
-L --language <lang>       Use <lang> instead of english
-p --platform <platform>   Use <platform> instead of the native platform
-a --auto-update-interval <days>
            Perform an automatic update if the cache is older than <days>
```

Or a much better example of the usage:

``` shellsession
$ tldr tldr
== tldr ==

Displays simple help pages for command-line tools, from the tldr-pages project.

More information: <https://tldr.sh>.

Get typical usages of a command (hint: this is how you got here!):
        tldr {{command}}

Show the tar tldr page for Linux:
        tldr -p {{linux}} {{tar}}

Get help for a git subcommand:
        tldr {{git-checkout}}

Update local pages (if the client supports caching):
        tldr -u
```

On the first run, this programme caches all available tldr pages. 
Since the number of available tldr pages rises quickly, it is recommended to regularly update the cache. 
Such an update can be run manually with `tldr --update`. 
Users of this client can enable automatic updates by running it with the option `--auto-update-interval DAYS` specified.
The client will then check whether the cached version of the tldr pages is older than `DAYS` days and perform an update in that case.
To enable this functionality permanently, users can put the line `alias tldr="tldr --auto-update-interval DAYS"` in their shell configuration file (e.g. `.bashrc`, `.zshrc`) with the desired update interval specified.

## Benchmarks

I compared fast-tldr to a few other implementations using the `benchmark.sh` script. These are my results:

```shellsession
$ ./benchmark.sh
Benchmark #1: tldr tar
  Time (mean ± σ):      48.7 ms ±   8.2 ms    [User: 3.9 ms, System: 3.6 ms]
  Range (min … max):    38.3 ms …  77.0 ms    23 runs
 
Benchmark #2: ./tealdeer tar
  Time (mean ± σ):      51.3 ms ±   7.9 ms    [User: 6.3 ms, System: 4.8 ms]
  Range (min … max):    43.3 ms …  74.3 ms    20 runs
 
Benchmark #3: ./tldr-c-client tar
  Time (mean ± σ):      95.6 ms ±  12.9 ms    [User: 6.6 ms, System: 11.3 ms]
  Range (min … max):    80.5 ms … 131.9 ms    15 runs
 
Benchmark #4: ./tldr-bash-client tar
  Time (mean ± σ):      97.9 ms ±   8.2 ms    [User: 29.6 ms, System: 12.3 ms]
  Range (min … max):    80.3 ms … 108.3 ms    17 runs
 
Benchmark #5: ./tldr-go-client tar
  Time (mean ± σ):     141.8 ms ±   6.9 ms    [User: 97.4 ms, System: 16.3 ms]
  Range (min … max):   133.2 ms … 156.0 ms    13 runs
 
Summary
  'tldr tar' ran
    1.05 ± 0.24 times faster than './tealdeer tar'
    1.96 ± 0.42 times faster than './tldr-c-client tar'
    2.01 ± 0.38 times faster than './tldr-bash-client tar'
    2.91 ± 0.51 times faster than './tldr-go-client tar'
```

Compared clients:
- [fast-tldr](https://github.com/gutjuri/fast-tldr/)
- [Tealdeer](https://github.com/dbrgn/tealdeer/)
- [C client](https://github.com/tldr-pages/tldr-cpp-client)
- [Bash client](https://github.com/pepa65/tldr-bash-client)
- [Go client by k3mist](https://github.com/k3mist/tldr/)

As we can see, fast-tldr is the fastest client (although not by a large margin).

CHANGELOG.md
# 1.0

* first release


test/Spec.hs
-- Copyright (C) 2020 Juri Dispan
--
-- This program is free software; you can redistribute it and/or modify it under
-- the terms of the GNU General Public License as published by the Free Software
-- Foundation; either version 2 of the License, or (at your option) any later
-- version.
--
-- This program is distributed in the hope that it will be useful, but WITHOUT
-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
-- FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along with
-- this program; if not, write to the Free Software Foundation, Inc., 59 Temple
-- Place, Suite 330, Boston, MA 02111-1307 USA

main :: IO ()
main = error "Not implemented yet."

app/Main.hs
-- Copyright (C) 2020 Juri Dispan
--
-- This program is free software; you can redistribute it and/or modify it under
-- the terms of the GNU General Public License as published by the Free Software
-- Foundation; either version 2 of the License, or (at your option) any later
-- version.
--
-- This program is distributed in the hope that it will be useful, but WITHOUT
-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
-- FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along with
-- this program; if not, write to the Free Software Foundation, Inc., 59 Temple
-- Place, Suite 330, Boston, MA 02111-1307 USA

module Main where

import           Tldr.App                       ( appMain )

main :: IO ()
main = appMain


src/Tldr.hs
-- Copyright (C) 2020 Juri Dispan
--
-- This program is free software; you can redistribute it and/or modify it under
-- the terms of the GNU General Public License as published by the Free Software
-- Foundation; either version 2 of the License, or (at your option) any later
-- version.
--
-- This program is distributed in the hope that it will be useful, but WITHOUT
-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
-- FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along with
-- this program; if not, write to the Free Software Foundation, Inc., 59 Temple
-- Place, Suite 330, Boston, MA 02111-1307 USA

{-# LANGUAGE OverloadedStrings #-}

module Tldr
  ( renderPage
  )
where

import           Data.Monoid                    ( (<>) )
import           Data.Text               hiding ( cons )
import           System.Console.ANSI
import qualified Data.Text                     as T
import qualified Data.Text.IO                  as TIO

renderHeading :: Text -> IO ()
renderHeading ln = do
  setSGR [SetColor Foreground Dull White]
  TIO.putStrLn $ "== " <> T.drop 2 ln <> " ==\n"

renderDescription :: Text -> IO ()
renderDescription ln = do
  setSGR [SetColor Foreground Dull White]
  TIO.putStrLn $ T.drop 2 ln <> "\n"

renderCommandDescription :: Text -> IO ()
renderCommandDescription ln = do
  setSGR [SetColor Foreground Dull Green]
  TIO.putStrLn $ T.drop 2 ln

renderCommand :: Text -> IO ()
renderCommand ln = do
  setSGR [SetColor Foreground Dull Blue]
  TIO.putStrLn $ "\t" <> T.dropEnd 1 (T.drop 1 ln) <> "\n"

renderLine :: Text -> IO ()
renderLine ln | "#" `T.isPrefixOf` ln = renderHeading ln
              | ">" `T.isPrefixOf` ln = renderDescription ln
              | "-" `T.isPrefixOf` ln = renderCommandDescription ln
              | "`" `T.isPrefixOf` ln = renderCommand ln
              | "" == ln              = return ()
              | otherwise             = error "Malformed TLDR-Page"

renderPage :: FilePath -> IO ()
renderPage fname = do
  text <- TIO.readFile fname
  mapM_ renderLine $ T.lines text


benchmark.sh
#!/bin/sh

# assumes fast-tldr is installed.

mkdir -p benchmarks && cd benchmarks

TESTPAGE=tar

# get tealdeer
if [ ! -f tealdeer ]; then
  wget https://github.com/dbrgn/tealdeer/releases/download/v1.4.1/tldr-linux-x86_64-musl
  mv tldr-linux-x86_64-musl tealdeer
  chmod +x tealdeer
  ./tealdeer --update
fi

if [ ! -f tldr-hs ]; then
  wget https://github.com/psibi/tldr-hs/releases/download/v0.8.0/tldr-linux-v0.8.0.tar.gz
  tar xvzf tldr-linux-v0.8.0.tar.gz
  mv tldr tldr-hs
  ./tldr-hs --update
fi

if [ ! -f tldr-c-client ]; then
  wget https://github.com/tldr-pages/tldr-c-client/archive/v1.3.0.zip
  unzip v1.3.0.zip
  cd tldr-c-client-1.3.0
  make
  mv ./tldr ../tldr-c-client
  cd ..
  ./tldr-c-client --update
fi

if [ ! -f tldr-bash-client ]; then
  wget -qO tldr-bash-client https://4e4.win/tldr
  sudo chmod +x tldr-bash-client
  ./tldr-bash-client update
fi

if [ ! -f tldr-go-client ]; then
  wget https://bitbucket.org/djr2/tldr/raw/f412bddb2a583ddd8192587bb33d0260bc4945d0/dist/linux.tar.bz2
  tar xvjf linux.tar.bz2
  mv dist/build/linux/tldr tldr-go-client
  chmod +x tldr-go-client
  ./tldr-go-client --update
fi


hyperfine --prepare 'sync; echo 3 | sudo tee /proc/sys/vm/drop_caches' \
          "tldr $TESTPAGE" \
          "./tealdeer $TESTPAGE" \
          "./tldr-hs $TESTPAGE" \
          "./tldr-c-client $TESTPAGE" \
          "./tldr-bash-client $TESTPAGE" \
          "./tldr-go-client $TESTPAGE"

src/Tldr/App.hs
-- Copyright (C) 2020 Juri Dispan
--
-- This program is free software; you can redistribute it and/or modify it under
-- the terms of the GNU General Public License as published by the Free Software
-- Foundation; either version 2 of the License, or (at your option) any later
-- version.
--
-- This program is distributed in the hope that it will be useful, but WITHOUT
-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
-- FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along with
-- this program; if not, write to the Free Software Foundation, Inc., 59 Temple
-- Place, Suite 330, Boston, MA 02111-1307 USA

module Tldr.App
  ( appMain
  )
where

import           Data.List                      ( isPrefixOf )
import           System.Environment             ( getArgs )
import           Tldr.App.Handler
import           Tldr.Types
import           Control.Monad
import           Text.Read

setIfView :: ViewOptions -> Command -> Command
setIfView (ViewOptions po lo aui) (ViewPage (ViewOptions po' lo' aui') ps) =
  ViewPage (ViewOptions (po `mplus` po') (lo `mplus` lo') (aui `mplus` aui')) ps
setIfView _ x = x

parseNoFlags :: [String] -> Command
parseNoFlags = ViewPage (ViewOptions Nothing Nothing Nothing)

addCommand :: String -> Command -> Command
addCommand p (ViewPage vo ps) = ViewPage vo (p : ps)
addCommand _ x                = x

parseOpts :: [String] -> Command
parseOpts (x : _) | x `elem` ["-u", "--update"] = UpdateIndex
parseOpts (x : _) | x `elem` ["-h", "--help"]   = ShowHelp 0
parseOpts (x : _) | x `elem` ["-v", "--version"]   = ShowVersion
parseOpts (x : y : xs) | x `elem` ["-L", "--language"] =
  setIfView (ViewOptions Nothing (Just y) Nothing) $ parseOpts xs
parseOpts (x : y : xs) | x `elem` ["-p", "--platform"] =
  setIfView (ViewOptions (Just y) Nothing Nothing) $ parseOpts xs
parseOpts (x : y : xs) | x `elem` ["-a", "--auto-update-interval"] =
  case readMaybe y of
    Nothing -> ShowHelp 1
    updateInterval@(Just _) ->
      setIfView (ViewOptions Nothing Nothing updateInterval) $ parseOpts xs
parseOpts ("--" : xs) = parseNoFlags xs
parseOpts (x : xs) | not ("-" `isPrefixOf` x) = addCommand x $ parseOpts xs
                   | otherwise                = ShowHelp 1
parseOpts [] = ViewPage (ViewOptions Nothing Nothing Nothing) []

appMain :: IO ()
appMain = do
  args <- getArgs
  let opts = parseOpts args
  handleTldrOpts opts


src/Tldr/Types.hs
-- Copyright (C) 2020 Juri Dispan
--
-- This program is free software; you can redistribute it and/or modify it under
-- the terms of the GNU General Public License as published by the Free Software
-- Foundation; either version 2 of the License, or (at your option) any later
-- version.
--
-- This program is distributed in the hope that it will be useful, but WITHOUT
-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
-- FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along with
-- this program; if not, write to the Free Software Foundation, Inc., 59 Temple
-- Place, Suite 330, Boston, MA 02111-1307 USA

module Tldr.Types where

data Command
  = UpdateIndex
  | ShowHelp Int
  | ShowVersion
  | ViewPage ViewOptions
             [String]
  deriving (Eq)

data ViewOptions =
  ViewOptions
    { platformOption :: Maybe String
    , languageOption :: Maybe String
    , autoUpdateInterval :: Maybe Int
    }
  deriving (Eq)


src/Tldr/App/Handler.hs
-- Copyright (C) 2020 Juri Dispan
--
-- This program is free software; you can redistribute it and/or modify it under
-- the terms of the GNU General Public License as published by the Free Software
-- Foundation; either version 2 of the License, or (at your option) any later
-- version.
--
-- This program is distributed in the hope that it will be useful, but WITHOUT
-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
-- FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along with
-- this program; if not, write to the Free Software Foundation, Inc., 59 Temple
-- Place, Suite 330, Boston, MA 02111-1307 USA

{-# LANGUAGE RecordWildCards #-}

module Tldr.App.Handler
  ( handleTldrOpts
  )
where

import           Data.Char                      ( toLower )
import           Data.List                      ( intercalate
                                                , delete
                                                , isPrefixOf
                                                )
import           Data.Semigroup                 ( (<>) )
import           Data.Time
import           Data.Maybe                     ( isJust )
import           Data.Version                   ( showVersion )
import           System.Directory
import           System.Exit                    ( exitFailure
                                                , exitSuccess
                                                )
import           System.FilePath                ( (<.>)
                                                , (</>)
                                                )
import           System.IO                      ( hPutStrLn
                                                , stderr
                                                )
import           Network.HTTP.Simple
import           Codec.Archive.Zip
import           Control.Monad

import           Tldr
import           Tldr.Types
import           Paths_tldr                     ( version )

tldrDirName :: String
tldrDirName = "tldr"

pagesUrl :: String
pagesUrl = "https://tldr.sh/assets/tldr.zip"

platformDirs :: [String]
platformDirs = ["linux", "osx", "windows", "sunos"]

showHelp :: Int -> IO ()
showHelp exitCode = do
  let
    helpMsg = unlines
      [ "Usage:"
      , "\ttldr [options] COMMAND"
      , "\ttldr [options]"
      , "Options:"
      , "-h --help                  Show help"
      , "-v,--version               Show version"
      , "-u --update                Update local cache"
      , "-L --language <lang>       Use <lang> instead of english"
      , "-p --platform <platform>   Use <platform> instead of the native platform"
      , "-a --auto-update-interval <days>"
      , "            Perform an automatic update if the cache is older than <days>"
      ]
  putStr helpMsg
  if exitCode == 0 then exitSuccess else exitFailure

showVersionInfo :: IO ()
showVersionInfo = do
  let versionMsg = unlines
        [ "fast-tldr " <> showVersion version
        , "Copyright (C) 2020 Juri Dispan, Sibi Prabakaran"
        ]
  putStr versionMsg

getLocaleSuffix :: Maybe String -> Maybe String
getLocaleSuffix lang = case map toLower <$> lang of
  Nothing  -> Nothing
  Just loc -> if "en" `isPrefixOf` loc then Nothing else Just loc

handleTldrOpts :: Command -> IO ()
handleTldrOpts command = case command of
  UpdateIndex             -> updateTldrPages
  ShowHelp exitCode       -> showHelp exitCode
  ShowVersion             -> showVersionInfo
  ViewPage _        []    -> handleTldrOpts $ ShowHelp 1
  ViewPage voptions pages -> do
    shouldPerformUpdate <- updateNecessary command
    when shouldPerformUpdate updateTldrPages
    let npage        = intercalate "-" pages
    let localeSuffix = getLocaleSuffix $ languageOption voptions
    fname <- getPagePath localeSuffix npage (getCheckDirs voptions)
    case fname of
      Just path -> renderPage path
      Nothing   -> if isJust localeSuffix
        then handleTldrOpts
          $ ViewPage voptions { languageOption = Nothing } pages
        else do
          hPutStrLn stderr ("No tldr entry for " <> unwords pages)
          exitFailure

updateNecessary :: Command -> IO Bool
updateNecessary (ViewPage ViewOptions {..} _) = do
  dataDir       <- getXdgDirectory XdgData tldrDirName
  dataDirExists <- doesDirectoryExist dataDir
  if not dataDirExists
    then return True
    else case autoUpdateInterval of
      Nothing             -> return False
      Just updateInterval -> do
        lastCachedTime <- getModificationTime dataDir
        currentTime    <- getCurrentTime
        let diffExceedsLimit =
              currentTime
                `diffUTCTime` lastCachedTime
                >             fromIntegral updateInterval
                *             nominalDay
        return diffExceedsLimit
updateNecessary _ = return False

updateTldrPages :: IO ()
updateTldrPages = do
  dataDir <- getXdgDirectory XdgData tldrDirName
  removePathForcibly dataDir
  createDirectory dataDir
  putStrLn $ "Downloading tldr pages to " ++ dataDir
  response <- httpLBS $ parseRequest_ pagesUrl
  let zipArchive = toArchive $ getResponseBody response
  extractFilesFromArchive [OptDestination dataDir] zipArchive

getPagePath :: Maybe String -> String -> [String] -> IO (Maybe FilePath)
getPagePath localeSuffix page pDirs = do
  dataDir <- getXdgDirectory XdgData tldrDirName
  let currentLocaleDir = case localeSuffix of
        Nothing -> "pages"
        Just xs -> "pages." <> xs
      pageDir = dataDir </> currentLocaleDir
      paths   = map (\x -> pageDir </> x </> page <.> "md") pDirs
  getFirstExisting paths

getFirstExisting :: [FilePath] -> IO (Maybe FilePath)
getFirstExisting []       = return Nothing
getFirstExisting (x : xs) = do
  exists <- doesFileExist x
  if exists then return $ Just x else getFirstExisting xs

getCheckDirs :: ViewOptions -> [String]
getCheckDirs voptions = case platformOption voptions of
  Nothing       -> "common" : platformDirs
  Just platform -> "common" : platform : delete platform platformDirs
